sdk/ai/Azure.AI.Inference/src/Telemetry/OpenTelemetryScope.cs

```diff
@@ -0,0 +1,247 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using System.Text.Json;
+using static Azure.AI.Inference.Telemetry.OpenTelemetryInternalConstants;
+using static Azure.AI.Inference.Telemetry.OpenTelemetryConstants;
+
+namespace Azure.AI.Inference.Telemetry
+{
+    internal class OpenTelemetryScope : IDisposable
+    {
+        /// <summary>
+        /// Activity source is used to save events and log the tags.
+        /// On the ApplicationInsights events are logged to traces table.
+        /// The tags are not logged, but are shown in the console.
+        /// </summary>
+        private static readonly ActivitySource s_chatSource = new ActivitySource(ActivityName);
+        /// <summary>
+        /// Add histograms to log metrics.
+        /// On the ApplicationInsights metrics are logged to customMetrics table.
+        /// </summary>
+        private static readonly Meter s_meter = new Meter(ActivityName);
+        private static readonly Histogram<double> s_duration = s_meter.CreateHistogram<double>(
+            GenAiClientOperationDurationMetricName, "s", "Measures GenAI operation duration.");
+        private static readonly Histogram<long> s_tokens = s_meter.CreateHistogram<long>(
+            GenAiClientTokenUsageMetricName, "{token}", "Measures the number of input and output token used.");
+
+        private Activity m_activity;
+        private Stopwatch m_duration;
+        private TagList m_commonTags;
+        private StreamingRecordedResponse m_recordedStreamingResponse;
+
+        /// <summary>
+        /// Return the content of the message.
+        /// Content accessor is implemented only on the exact class, but not on the abstract generated class.
+        /// </summary>
+        /// <param name="message"></param>
+        /// <returns></returns>
+        /// <exception cref="NotSupportedException"></exception>
+        public static string GetContent(ChatRequestMessage message)
+        {
+            if (message.GetType() == typeof(ChatRequestAssistantMessage))
+                return ((ChatRequestAssistantMessage)message).Content;
+            if (message.GetType() == typeof(ChatRequestSystemMessage))
+                return ((ChatRequestSystemMessage)message).Content;
+            if (message.GetType() == typeof(ChatRequestToolMessage))
+                return ((ChatRequestToolMessage)message).Content;
+            if (message.GetType() == typeof(ChatRequestUserMessage))
+                return ((ChatRequestUserMessage)message).Content;
+            throw new NotSupportedException("The message type is not supported");
```

Comment:
telemetry code is not expected to throw, let's just ignore unknown types.

Reply:
Fixed.